vmsyscall/src/memory_map.rs

```diff
@@ -0,0 +1,317 @@
+        }
+    }
+
+    /// Peek the next index value.
+    fn next_entry_index(&self) -> usize {
+        self.next_entry_index as usize
+    }
+}
+
+impl Default for MemoryMap {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Deref for MemoryMap {
+    type Target = [MemoryRegion];
+
+    fn deref(&self) -> &Self::Target {
+        &self.entries[0..self.next_entry_index()]
+    }
+}
+
+impl DerefMut for MemoryMap {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        let next_index = self.next_entry_index();
+        &mut self.entries[0..next_index]
+    }
+}
+
+impl fmt::Debug for MemoryMap {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_list().entries(self.iter()).finish()
+    }
+}
+
+/// Represents a region of physical memory.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(C)]
+pub struct MemoryRegion {
+    /// The range of frames that belong to the region.
+    pub range: FrameRange,
+    /// The type of the region.
+    pub region_type: MemoryRegionType,
+}
+
+impl MemoryRegion {
+    /// Produce an empty `MemoryRegion`.
+    pub const fn empty() -> Self {
+        MemoryRegion {
+            range: FrameRange {
+                start_frame_number: 0,
+                end_frame_number: 0,
+            },
+            region_type: MemoryRegionType::Empty,
+        }
+    }
+}
+
+/// A range of frames with an exclusive upper bound.
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[repr(C)]
+pub struct FrameRange {
+    /// The frame _number_ of the first 4KiB frame in the region.
+    ///
+    /// This convert this frame number to a physical address, multiply it with the
+    /// page size (4KiB).
+    pub start_frame_number: u64,
+    /// The frame _number_ of the first 4KiB frame that does no longer belong to the region.
+    ///
+    /// This convert this frame number to a physical address, multiply it with the
+    /// page size (4KiB).
+    pub end_frame_number: u64,
+}
+
+impl FrameRange {
+    /// Create a new FrameRange from the passed start_addr and end_addr.
+    ///
+    /// The end_addr is exclusive.
+    pub fn new(start_addr: u64, end_addr: u64) -> Self {
+        let last_byte = end_addr - 1;
+        FrameRange {
+            start_frame_number: start_addr / PAGE_SIZE,
+            end_frame_number: (last_byte / PAGE_SIZE) + 1,
+        }
+    }
+
+    /// Returns true if the frame range contains no frames.
+    pub fn is_empty(&self) -> bool {
+        self.start_frame_number == self.end_frame_number
+    }
+
+    /// Length of the frame range
+    pub fn len(&self) -> u64 {
+        self.end_frame_number - self.start_frame_number
+    }
+
+    /// Returns the physical start address of the memory region.
+    pub fn start_addr(&self) -> u64 {
+        self.start_frame_number * PAGE_SIZE
+    }
+
+    /// Returns the physical end address of the memory region.
+    pub fn end_addr(&self) -> u64 {
+        self.end_frame_number * PAGE_SIZE
+    }
+}
```

Comment:
I think we should use `core::ops::Range` here instead. This is also isomorphic with `span::Span`. Make a newtype `Frame(u64)` with conversions to/from addresses. Then just do `Range<Frame>` or `Span<Frame>`.